### PR TITLE
Magic links and UI fixes

### PR DESCRIPTION
**Pull request by shivanmodha**
## Updates
 - Implemented `/magic` page to handle generic magic urls, with a special rule for magic urls starting with `forgot-`.
- Implemented `/forgot` so that users who forgot their password can reset it
 - Styling updates from @dhruvilp 
> For dev, someone set the expiration for forgot magic links to 3 seconds -___-

## Bug Fixes
- Fixed the footer issue that is messing with the width and height of the page, as can be seen at `dev.hackru.org`
    - Added to `defaults.js` so that its unified
image

- Rerouted assets from `./` to `/`
> WHY WAS THE FOOTER SO ANNOYING????

### CODE REVIEW

File: public/index.css [hunk 1]
```diff
@@ -188,4 +189,8 @@ h4{
     transform: skew(0deg, 4deg);
 }
 
+.table-hover:hover{
+    color: rgba(95, 255, 220, 1) !important;
+    text-decoration: none;
+}
```
Comment (thekabistro) [inline public/index.css]: This hover can be removed...It doesn't look very good and the table already has a hover effect.

File: public/index.css [hunk 2]
```diff
@@ -35,7 +35,7 @@ a:link:hover {
 }
 
 a:visited {
-    color: rgba(38, 232, 184, 1);
+    color: rgba(95,255,220, 0.8);
 }
```
Comment (thekabistro) [inline public/index.css]: We should set a couple color variables for the file and all these attributes should use the same colors.  Is there a good reason for the opacity to be 0.8 on visited links?

File: src/components/Landing/Landing.jsx [hunk 3]
```diff
@@ -29,9 +28,15 @@ class LandingPage extends Component {
...
+                if (i !== keys.length - 1) {
+                    className = "bg-no-gradient skew-right";
+                    if (i % 2 === 0) {
+                        className = "bg-gradient-right skew-right";
+                    }
+                } else {
+                    // The footer is a special type of human being...
+                    sectionClasses = "col-lg-12 offset-lg-0 col-xs-12 offset-xs-0 color-priority";
                 }
                 let style = {};
```
Comment (thekabistro) [inline src/components/Landing/Landing.jsx]: Ah.  I like this.

File: src/components/Landing/Sections/FaqsCollapse.jsx [hunk 4]
```diff
@@ -24,7 +24,7 @@ class FaqsCollapse extends React.Component {
                     <h5 className="" onClick={this.toggle}>{cat.title} <Icon className="faq-hover pull-right" name={(this.state.collapse) ? ("chevron-up") : ("chevron-down")} /></h5>
                 </strong>
                 <hr className="faq-hover"/>
-                <Collapse className="mb-3" isOpen={this.state.collapse}>{cat.text}</Collapse>
+                <Collapse isOpen={this.state.collapse}>{cat.text}<div style={{ minHeight: 25 }} /></Collapse>
```
Comment (thekabistro) [inline src/components/Landing/Sections/FaqsCollapse.jsx]: Why the change here?

File: src/components/Landing/Sections/Footer.jsx [hunk 5]
```diff
@@ -13,33 +13,31 @@ class Footer extends Component {
...
-                                <div className="col-12 mb-5 mb-lg-0 text-center col-lg-4 mb-4 align-middle">
-                                    <a style={{marginRight: 10 }} href="https://mlh.io/" target="_blank" rel="noopener noreferrer">
-                                        <img className="footer-logo" src="https://static.mlh.io/brand-assets/logo/official/mlh-logo-white.png" alt="MLH logo" />
-                                    </a>
-                                    <a style={{marginRight: 10 }} href="http://usacs.rutgers.edu/" target="_blank" rel="noopener noreferrer">
-                                        <img className="footer-logo" src="./assets/icons/usacs-logo.svg" alt="USACS logo" />
-                                    </a>
-                                </div>
+                <Row>
+                    <div className="footer-border bg-no-gradient"></div>
```
Comment (thekabistro) [inline src/components/Landing/Sections/Footer.jsx]: With the removal of the color from the previous section this footer-border div no longer needs to exist

File: src/components/Landing/Sections/Footer.jsx [hunk 6]
```diff
@@ -13,33 +13,31 @@ class Footer extends Component {
...
+                    <div className="footer" style={{ zIndex: 15, height: "100%", width: "100%" }}>
```
Comment (thekabistro) [inline src/components/Landing/Sections/Footer.jsx]: But I think that the custom width and height here will mess things up

File: src/components/Landing/Sections/Schedule.jsx [hunk 7]
```diff
@@ -26,7 +26,7 @@ class Schedule extends Component {
 
                 <div className="row row-content-box">
                     <div className="col-12 col-lg-6">
-                        <h3 className="text-center mb-2">Saturday</h3>
+                        <h3 className="text-center mb-2" style={{color: "#F8A5FF"}}>Saturday</h3>
                         <table className="table table-hover">
```
Comment (thekabistro) [inline src/components/Landing/Sections/Schedule.jsx]: I'm not a fan of the purple.  The green might be a better idea.  Anyways I would like us to get away from inline color changes.  These become impossible to maintain.

File: src/components/Landing/Sections/Schedule.jsx [hunk 8]
```diff
@@ -66,7 +66,7 @@ class Schedule extends Component {
                     </div>
 
                     <div className="col-12 mt-5 mt-lg-0 col-lg-6">
-                        <h3 className="text-center mb-2">Sunday</h3>
+                        <h3 className="text-center mb-2" style={{color: "#F8A5FF"}}>Sunday</h3>
```
Comment (thekabistro) [inline src/components/Landing/Sections/Schedule.jsx]: See above.

File: src/components/Profile.js [hunk 9]
```diff
@@ -269,6 +293,70 @@ class Profile {
...
+                        callback();
+                    } else {
+                        callback(body.body);
+                    }
+                }
+            });
+        }
+    }
 }
 /***************************************************************PROFILE***************************************************************/
```
Comment (thekabistro) [inline src/components/Profile.js]: These error checks are very trusting of LCS to always give you one of the codes that you're expecting.  Under the login case, I've managed to get some wacky unicode error, which does not have a status code returned, and does not return a callback of undefined error.  It'd be good if there's one more else statement for all these conditions that say unexpected error!

File: src/App.jsx [hunk 10]
```diff
@@ -68,16 +105,19 @@ class App extends Component {
...
+                        <Route exact path="/login" render={(props) => <LoginPage {...props} {...componentProps} />} />
+                        <Route exact path="/signup" render={(props) => <SignUpPage {...props} {...componentProps} />} />
                         <Route exact path="/logout" component={() => { this.state.profile.Logout(); this.setState({ profile: this.state.profile, loggedout: true }); return (<Redirect to="/" />); }} />
-                        <Route exact path="/dashboard" render={(props) => <DashboardPage {...props} isMobile={this.state.isMobile} profile={this.state.profile} />} />
+                        <Route exact path="/forgot" render={(props) => <ForgotPage {...props} {...componentProps} />} />
+                        <Route exact path="/magic/:mlurl" render={(props) => <MagicPage {...props} {...componentProps} />} />
+                        <Route exact path="/dashboard" render={(props) => <DashboardPage {...props} {...componentProps} />} />
+                        <Route exact path="/live" render={(props) => <Redirect to="/" />} /> {/* We will be implementing this in the future */}
```
Comment (thekabistro) [inline src/App.jsx]: Does it serve any purpose to have this route in here right now?  Yes this is what we are intending to do, but I don't know that it makes sense to include it until we have a page ready to go?

File: public/index.css (hunk 1 above)
Comment (shivanmodha) [reply public/index.css]: agreed

File: public/index.css (hunk 2 above)
Comment (shivanmodha) [reply public/index.css]: @dhruvilp I think thats to differentiate between the regular link?
```css
a:link {
    color: rgba(95,255,220, 1);
    transition: color 0.5s ease;
}
```

File: src/components/Landing/Sections/Footer.jsx (hunk 5 above)
Comment (shivanmodha) [reply src/components/Landing/Sections/Footer.jsx]: image

File: src/components/Landing/Sections/Footer.jsx (hunk 6 above)
Comment (shivanmodha) [reply src/components/Landing/Sections/Footer.jsx]: ^^ above

**Comment by shivanmodha** [comment]
Waiting on https://github.com/HackRU/lcs/pull/49

File: src/components/Landing/Sections/Footer.jsx [hunk 11]
```diff
@@ -14,8 +14,7 @@ class Footer extends Component {
         return (
             <div>
                 <Row>
-                    <div className="footer-border bg-no-gradient"></div>
-                    <div className="footer" style={{ zIndex: 15, height: "100%", width: "100%" }}>
+                    <div className="footer bg-no-gradient" style={{zIndex: 15}}>
                         <Row style={{ marginRight: 0, textAlign: "center" }}>
```
Comment (thekabistro) [inline src/components/Landing/Sections/Footer.jsx]: @shivanmodha in case you were curious

Comment (shivanmodha) [reply src/components/Landing/Sections/Footer.jsx]: awesome!

File: src/Defaults.js [hunk 12]
```diff
@@ -71,10 +71,19 @@ const navlinks = {
...
 /****************************************************************THEME****************************************************************/
+// To unify styling between javascript and css, we define these variables in "index.css" as a variable, and proceed to get a handle
+// to the respective values through JavaScript
+let computedStyle = getComputedStyle(document.body);
 const theme = {
-    "primary": ["#26E8BD", "#5FFFDC"],
-    "secondary": ["#5A7A96", "#354a5f"],
-    "accent": ["#FF80CF", "#F8A5FF"]
+    "primary": [
+        computedStyle.getPropertyValue("--primary-color"),
```
Comment (shivanmodha) [inline src/Defaults.js]: @thekabistro Now we won't have to redefine our theme separately in the CSS and JS. I think if we expand on this in the future and it becomes a lot, we should look to create a theme.css file

File: src/components/Profile.js (hunk 9 above)
Comment (shivanmodha) [reply src/components/Profile.js]: The else logic is already built in as follows:
```...
} else if (body.statusCode === 200) {
	callback();
} else {
	callback(body.body);
}
```
I'm returning the `body` string that LCS should send. Should I replace that to `Unexpected error` instead?

**Comment by shivanmodha** [comment]
## Updates
- Generic Magic Link support. For now, any magic links with prefix `forgot-` will trigger a password reset, while all other magic links will be directed to a login screen. After this, we have the groundwork for handling the other magic links that we get
- Responsive footer fix for both desktop and mobile
    - Hopefully, we don't have to deal with this again...
- Standard theme styling between CSS/JS to make things easier. We should work towards migrating some other attributes to this style to make things easier to edit/maintain in the future

File: src/components/Profile.js (hunk 9 above)
Comment (thekabistro) [reply src/components/Profile.js]: Yes... I've managed the following which contains no body string.  So that'd be one way to handle it. 

screen shot 2019-01-03 at 10 35 39 am

File: public/index.css [hunk 13]
```diff
@@ -1,87 +1,72 @@
 @import url('https://fonts.googleapis.com/css?family=Raleway+Dots');
 
+:root {
+    --primary-color: #26E8BD;
+    --primary-color-alt: #5FFFDC;
+    --secondary-color: #5A7A96;
+    --secondary-color-alt: #354a5f;
+    --accent-color: #FF80CF;
+    --accent-color-alt: #F8A5FF;
```
Comment (ryanisaacg) [inline public/index.css]: 👍

File: src/components/Forgot.jsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+            errors: ""
+        });
+    }
+    render() {
+        let innerText = "Password reset";
+        let innerForm = (
+            <div>
+                <FormGroup row>
+                    <InputGroup>
+                        <Input required type="email" id="email" placeholder="email" style={{ borderRadius: 0, background: "rgba(255, 255, 255, 0.2)", border: "none", color: "white" }} />
```
Comment (ryanisaacg) [inline src/components/Forgot.jsx]: I noticed you have what looks like the same CSS inline on every input element. Wouldn't it be better to make this a CSS rule in public/index.css?

File: src/App.jsx [hunk 15]
```diff
@@ -51,13 +56,45 @@ class App extends Component {
...
+            magic: ""
+        });
+    }
+    /**
+     * Set the application magic link
+     * @param {String} magic Magic link from lcs 
+     */
+    setMagic(magic) {
+        this.setState({
+            magic: magic
```
Comment (ryanisaacg) [inline src/App.jsx]: Nitpick: this can just be `magic`, as in `this.setState({ magic })`

**Comment by shivanmodha** [comment]
image

New background! Also, decreased aliens for mobile rendering purposes

**Comment by shivanmodha** [comment]
Darker background overlay!
image
image
image

File: public/index.css [hunk 16]
```diff
@@ -1,136 +1,69 @@
...
+    --input-border-radius: 0px;
+    --select-background: rgba(255, 255, 255, 0.75);
+    --select-color: rgba(0, 0, 0, 1);
+    --select-input-background: rgba(255, 255, 255, 0.075);
+    /* Hero Image */
+    --hero-width: 400px;
+    --hero-height: 400px;
+    --hero-border-radius: 35%;
+    --hero-background: radial-gradient(rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0));
+}
```
Comment (shivanmodha) [inline public/index.css]: @thekabistro Most inline theme styles are now controlled by this! One stop shop

File: src/components/Profile.js [hunk 17]
```diff
@@ -14,7 +14,7 @@ import { defaults } from "../Defaults";
 /**
  * Configure the all of the urls that we will need to access the rest api
  */
-const BASE = (process.env.NODE_ENV && process.env.NODE_ENV === "development") ? (defaults.rest.dev) : (defaults.rest.prod);
+const BASE = ((process.env.NODE_ENV && process.env.NODE_ENV === "development") || window.location.origin.includes("dev.hackru.org")) ? (defaults.rest.dev) : (defaults.rest.prod);
```
Comment (shivanmodha) [inline src/components/Profile.js]: hopefully this will make it easier for ya'll to deploy

File: public/index.css [hunk 18]
```diff
@@ -197,4 +118,73 @@ h4{
...
+    text-align: center;
+    font-size: 1.25rem;
+    font-weight: bold;
+}
+#landing-section {
+    color: white;
+}
 
+/* Form Inputs (Dashboard/Login/Signup/Magic/Forgot) */
+input, textarea, .custom-file-label {
```
Comment (shivanmodha) [inline public/index.css]: We changed the input styles anyway, so now its the default for inputs

File: src/Defaults.js [hunk 19]
```diff
@@ -30,33 +30,75 @@ const navlinks = {
...
+    "input-border-radius",
+    "select-background",
+    "select-color",
+    "select-input-background",
+    // Hero Image
+    "hero-width",
+    "hero-height",
+    "hero-border-radius",
+    "hero-background"
+];
```
Comment (shivanmodha) [inline src/Defaults.js]: If there's a blanket function to get all CSS variables, we wont have to populate this manually in the future

File: public/index.css (hunk 16 above)
Comment (thekabistro) [reply public/index.css]: I love it!!
